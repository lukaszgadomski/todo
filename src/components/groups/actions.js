export const types = {
  ADD_GROUP: "ADD_GROUP",
  EDIT_GROUP: "EDIT_GROUP",
  DELETE_GROUP: "DELETE_GROUP",
  SEARCH_GROUP: "SEARCH_GROUP",
  SELECT_GROUP: "SELECT_GROUP",
  CLOSE_GROUP: "CLOSE_GROUP",

  ADD_TODO: "ADD_TODO",
  EDIT_TODO: "EDIT_TODO",
  COMPLETE_TODO: "COMPLETE_TODO",
  DELETE_TODO: "DELETE_TODO"
};

export const addGroup = text => ({ type: types.ADD_GROUP, text });
export const editGroup = (id, text) => ({ type: types.EDIT_GROUP, id, text });
export const deleteGroup = id => ({ type: types.DELETE_GROUP, id });
export const searchGroup = text => ({
  type: types.SEARCH_GROUP,
  text
});
export const selectGroup = id => ({ type: types.SELECT_GROUP, id });
export const closeGroup = () => ({ type: types.CLOSE_GROUP });

export const addTodo = text => ({
  type: types.ADD_TODO,
  text
});
export const editTodo = (id, text) => ({
  type: types.EDIT_TODO,
  id,
  text
});
export const completeTodo = id => ({
  type: types.COMPLETE_TODO,
  id
});
export const deleteTodo = id => ({
  type: types.DELETE_TODO,
  id
});
