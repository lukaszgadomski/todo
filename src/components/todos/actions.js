export const types = {
  SEARCH_TODO: "SEARCH_TODO",
  CHANGE_FILTER_TODO: "CHANGE_FILTER_TODO"
};

export const searchTodo = text => ({
  type: types.SEARCH_TODO,
  text
});
export const filterTodo = filter => ({
  type: types.CHANGE_FILTER_TODO,
  filter
});
