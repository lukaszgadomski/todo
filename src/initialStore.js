export const filterTodoType = {
  ALL: "all",
  ACTIVE: "incompleted",
  COMPLETED: "completed"
};

export default {
  //group component state
  groups: {
    search: "",
    list: [],
    todos_by_group_id: {}, //list of todos for each list by group id
    selected_group_id: null,
    visible: true
  },
  //todos component state
  todos: {
    search: "",
    list: null, //todos from selected group, if it equals null then no group is selected
    filter: filterTodoType.ALL
  }
};
