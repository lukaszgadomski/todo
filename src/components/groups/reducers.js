import initialState, { filterTodoType } from "../../initialStore";
import { types } from "./actions";

import add from "ramda/src/add";
import evolve from "ramda/src/evolve";
import always from "ramda/src/always";
import append from "ramda/src/append";
import map from "ramda/src/map";
import curry from "ramda/src/curry";
import filter from "ramda/src/filter";
import assoc from "ramda/src/assoc";
import cond from "ramda/src/cond";
import propEq from "ramda/src/propEq";
import T from "ramda/src/T";
import dissoc from "ramda/src/dissoc";

export const getMaxId = curry((field, list) => {
  return list.reduce((maxId, item) => Math.max(item[field], maxId), 0) + 1;
});

const editById = curry((id, fn) =>
  map(
    cond([
      [propEq("id", id), fn],
      [T, item => item] //return original item if cond returns false
    ])
  )
);

const appendWithId = data => list => {
  if (data === undefined) {
    throw new Error("data undefined");
  }
  if (list === undefined) {
    throw new Error("empty list");
  }
  return append(assoc("id", getMaxId("id", list), data), list);
};

export default (state = initialState.groups, action) => {
  let groupId = null;

  switch (action.type) {
    case types.ADD_GROUP:
      const id = getMaxId("id", state.list);
      return evolve({
        list: append({
          id: id,
          name: action.text,
          count: 0
        }),
        todos_by_group_id: assoc(id, [])
      })(state);
    case types.EDIT_GROUP:
      return evolve({
        list: editById(action.id, assoc("name", action.text))
      })(state);
    case types.DELETE_GROUP:
      return evolve({
        list: filter(item => item.id !== action.id),
        todos_by_group_id: dissoc(action.id)
      })(state);
    case types.SEARCH_GROUP:
      return evolve({
        search: always(action.text)
      })(state);
    case types.SELECT_GROUP:
      return evolve({
        visible: always(false),
        selected_group_id: always(action.id)
      })(state);
    case types.CLOSE_GROUP:
      return evolve({
        visible: always(true),
        selected_group_id: always(null)
      })(state);
    case types.ADD_TODO:
      groupId = state.selected_group_id;
      return evolve({
        list: editById(
          groupId,
          evolve({
            count: add(1)
          })
        ),
        todos_by_group_id: evolve({
          [groupId]: appendWithId({
            parent_id: groupId,
            name: action.text,
            state: filterTodoType.ACTIVE
          })
        })
      })(state);
    case types.EDIT_TODO:
      groupId = state.selected_group_id;
      return evolve({
        todos_by_group_id: evolve({
          [groupId]: editById(action.id, assoc("name", action.text))
        })
      })(state);
    case types.COMPLETE_TODO:
      groupId = state.selected_group_id;
      return evolve({
        todos_by_group_id: evolve({
          [groupId]: editById(
            action.id,
            cond([
              [
                propEq("state", filterTodoType.COMPLETED),
                assoc("state", filterTodoType.ACTIVE)
              ],
              [
                propEq("state", filterTodoType.ACTIVE),
                assoc("state", filterTodoType.COMPLETED)
              ],
              [T, s => s] //return original item if cond returns false
            ])
          )
        })
      })(state);
    case types.DELETE_TODO:
      groupId = state.selected_group_id;
      return evolve({
        list: editById(
          groupId,
          evolve({
            count: add(-1)
          })
        ),
        todos_by_group_id: evolve({
          [groupId]: filter(item => item.id !== action.id)
        })
      })(state);
    default:
      return state;
  }
};
