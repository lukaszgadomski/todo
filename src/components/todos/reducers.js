import initialState from "../../initialStore";
import { types } from "./actions";

import evolve from "ramda/src/evolve";
import always from "ramda/src/always";

export default (state = initialState.todos, action) => {
  switch (action.type) {
    case types.SEARCH_TODO:
      return evolve({
        search: always(action.text)
      })(state);
    case types.CHANGE_FILTER_TODO:
      return evolve({
        filter: always(action.filter)
      })(state);
    default:
      return state;
  }
};
