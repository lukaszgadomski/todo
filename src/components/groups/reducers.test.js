import reducers from "./reducers";
import * as actions from "./actions";
import initialStore, { filterTodoType } from "../../initialStore";

const stringy = obj => JSON.stringify(obj, null, 4);

let state = {};
describe("reducers", () => {
  describe("group", () => {
    beforeEach(() => {
      state = {
        ...initialStore
      };
    });
    test("add", () => {
      expect(
        reducers(state.groups, actions.addGroup("helou"))
      ).toMatchSnapshot();
    });
    test("edit", () => {
      state.groups = reducers(state.groups, actions.addGroup("helou"));
      expect(
        reducers(state.groups, actions.editGroup(1, "helouedited"))
      ).toMatchSnapshot();
    });
    test("delete", () => {
      state.groups = reducers(state.groups, actions.addGroup("helou"));
      state.groups = reducers(state.groups, actions.addGroup("helou 2"));
      expect(reducers(state.groups, actions.deleteGroup(1))).toMatchSnapshot();
    });
    test("search", () => {
      state.groups = reducers(state.groups, actions.addGroup("helou"));
      state.groups = reducers(
        state.groups,
        actions.addGroup("helou are you there?")
      );
      expect(
        reducers(state.groups, actions.searchGroup("you"))
      ).toMatchSnapshot();
    });
    test("select", () => {
      state.groups = reducers(state.groups, actions.addGroup("helou"));
      expect(reducers(state.groups, actions.selectGroup(1))).toMatchSnapshot();
    });
    test("close", () => {
      state.groups.selected_group_id = 111;
      expect(reducers(state.groups, actions.closeGroup())).toMatchSnapshot();
    });
  });

  describe("todo", () => {
    beforeEach(() => {
      state = {
        ...initialStore
      };
      state.groups = reducers(state.groups, actions.addGroup("nr 1"));

      state.groups = reducers(state.groups, actions.addGroup("nr 2"));
      state.groups = reducers(state.groups, actions.selectGroup(2));
      state.groups = reducers(state.groups, actions.addTodo("helou", 2));
      state.groups = reducers(state.groups, actions.addTodo("helou 2", 2));
      state.groups = reducers(state.groups, actions.selectGroup(1));
    });
    test("add", () => {
      expect(
        reducers(state.groups, actions.addTodo("helou", 1))
      ).toMatchSnapshot();
    });
    test("edit", () => {
      state.groups = reducers(state.groups, actions.addTodo("helou", 1));
      expect(
        reducers(state.groups, actions.editTodo(1, 1, "edited"))
      ).toMatchSnapshot();
    });
    test("delete", () => {
      state.groups = reducers(state.groups, actions.addTodo("helou", 1));
      state.groups = reducers(state.groups, actions.addTodo("helou 2", 1));
      state.groups = reducers(state.groups, actions.addTodo("helou 3", 1));
      expect(
        reducers(state.groups, actions.deleteTodo(1, 1))
      ).toMatchSnapshot();
    });
  });
});
