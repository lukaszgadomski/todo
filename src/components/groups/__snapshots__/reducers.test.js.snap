// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers group add 1`] = `
Object {
  "list": Array [
    Object {
      "count": 0,
      "id": 1,
      "name": "helou",
    },
  ],
  "search": "",
  "selected_group_id": null,
  "todos_by_group_id": Object {
    "1": Array [],
  },
  "visible": true,
}
`;

exports[`reducers group close 1`] = `
Object {
  "list": Array [],
  "search": "",
  "selected_group_id": null,
  "todos_by_group_id": Object {},
  "visible": true,
}
`;

exports[`reducers group delete 1`] = `
Object {
  "list": Array [
    Object {
      "count": 0,
      "id": 2,
      "name": "helou 2",
    },
  ],
  "search": "",
  "selected_group_id": null,
  "todos_by_group_id": Object {
    "2": Array [],
  },
  "visible": true,
}
`;

exports[`reducers group edit 1`] = `
Object {
  "list": Array [
    Object {
      "count": 0,
      "id": 1,
      "name": "helouedited",
    },
  ],
  "search": "",
  "selected_group_id": null,
  "todos_by_group_id": Object {
    "1": Array [],
  },
  "visible": true,
}
`;

exports[`reducers group search 1`] = `
Object {
  "list": Array [
    Object {
      "count": 0,
      "id": 1,
      "name": "helou",
    },
    Object {
      "count": 0,
      "id": 2,
      "name": "helou are you there?",
    },
  ],
  "search": "you",
  "selected_group_id": null,
  "todos_by_group_id": Object {
    "1": Array [],
    "2": Array [],
  },
  "visible": true,
}
`;

exports[`reducers group select 1`] = `
Object {
  "list": Array [
    Object {
      "count": 0,
      "id": 1,
      "name": "helou",
    },
  ],
  "search": "",
  "selected_group_id": 1,
  "todos_by_group_id": Object {
    "1": Array [],
  },
  "visible": false,
}
`;

exports[`reducers todo add 1`] = `
Object {
  "list": Array [
    Object {
      "count": 1,
      "id": 1,
      "name": "nr 1",
    },
    Object {
      "count": 2,
      "id": 2,
      "name": "nr 2",
    },
  ],
  "search": "",
  "selected_group_id": 1,
  "todos_by_group_id": Object {
    "1": Array [
      Object {
        "id": 1,
        "name": "helou",
        "parent_id": 1,
        "state": "active",
      },
    ],
    "2": Array [
      Object {
        "id": 1,
        "name": "helou",
        "parent_id": 2,
        "state": "active",
      },
      Object {
        "id": 2,
        "name": "helou 2",
        "parent_id": 2,
        "state": "active",
      },
    ],
  },
  "visible": false,
}
`;

exports[`reducers todo delete 1`] = `
Object {
  "list": Array [
    Object {
      "count": 2,
      "id": 1,
      "name": "nr 1",
    },
    Object {
      "count": 2,
      "id": 2,
      "name": "nr 2",
    },
  ],
  "search": "",
  "selected_group_id": 1,
  "todos_by_group_id": Object {
    "1": Array [
      Object {
        "id": 2,
        "name": "helou 2",
        "parent_id": 1,
        "state": "active",
      },
      Object {
        "id": 3,
        "name": "helou 3",
        "parent_id": 1,
        "state": "active",
      },
    ],
    "2": Array [
      Object {
        "id": 1,
        "name": "helou",
        "parent_id": 2,
        "state": "active",
      },
      Object {
        "id": 2,
        "name": "helou 2",
        "parent_id": 2,
        "state": "active",
      },
    ],
  },
  "visible": false,
}
`;

exports[`reducers todo edit 1`] = `
Object {
  "list": Array [
    Object {
      "count": 1,
      "id": 1,
      "name": "nr 1",
    },
    Object {
      "count": 2,
      "id": 2,
      "name": "nr 2",
    },
  ],
  "search": "",
  "selected_group_id": 1,
  "todos_by_group_id": Object {
    "1": Array [
      Object {
        "id": 1,
        "name": "edited",
        "parent_id": 1,
        "state": "active",
      },
    ],
    "2": Array [
      Object {
        "id": 1,
        "name": "helou",
        "parent_id": 2,
        "state": "active",
      },
      Object {
        "id": 2,
        "name": "helou 2",
        "parent_id": 2,
        "state": "active",
      },
    ],
  },
  "visible": false,
}
`;
